File: number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(i,j):
            for dirs in directions:
                r=i+dirs[0]
                c=j+dirs[1]
                if r>=0 and r<len(grid) and c>=0 and c<len(grid[0]) and grid[r][c]=="1":
                    grid[r][c]="#"
                    dfs(r,c)
        count=0
        directions=((0,1),(1,0),(-1,0),(0,-1))
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]=="1":
                    grid[i][j]="#"
                    dfs(i,j)
                    count+=1
​
        return count
                    
                    
        

File: number-of-islands/number-of-islands.java
class Solution {

    public void dfs(int i, int j, int[][] dirs, int n, int m, char[][] grid){
        for (int[] d:dirs){
            int r=i+d[0];
            int c=j+d[1];
            if (r>=0 && r<n && c>=0 && c<m && grid[r][c]=='1'){
                grid[r][c]='#';
                dfs(r,c,dirs,n,m,grid);
            }
            
        }
    }
    public int numIslands(char[][] grid) {
        int count=0;
        int[][] dirs = new int[][]{{1,0},{0,1},{0,-1},{-1,0}};
        int n=grid.length;
        int m=grid[0].length;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if (grid[i][j]=='1'){
                    grid[i][j]='#';
                    dfs(i,j,dirs,n,m,grid);
                    count++;
                    
            }
            }
        }
        return count;
        
        
    }
}